### DISCUSSION TITLE
How to handle redirects with additional segments before locale?

### DISCUSSION THREAD

**Discussion by Maclay74** [amannn/next-intl#1547] [Q&A]
Hello.

I'm working on a website that needs to have a region segment before locale and handle redirects accordingly. 
Like `/page` or `/en/page/`  should lead to `/us/en/page`. 

It might work with new custom prefixes, but the problem is, we get region from the web server, so.. we need to somehow set the correct one. Moreover, region doesn't affect our language, we can have `/es/es-mx` (don't ask, I know, I know).

So, how should I handle redirects in this case? I am looking at [this docs section](https://next-intl-docs.vercel.app/docs/routing/middleware#example-additional-rewrites) but can't figure out how to handle it in links, for example.

**Comment by HappyEmu** [comment]
I'd love to know that as well. I wanted to move an existing app over to NextJS and its URL structure comprises a [country] dynamic segment before the [locale] segment. I see no way to do that at the moment.

**Reply by Maclay74** [reply]
Hey, we found a solution for that. It works for us, although it's a bit clumsy..

So our middleware has two redirects that make sure that eventually user is on `/country/locale` path. 
Then, in middleware still, we split this url and overwrite nextUrl, removing the region. 

Kinda like 
```ts
// redirects until we got /country/locale/page
const [region, locale, ...rest] = request.nextUrl.pathname;
request.nextUrl.pathname = ["", locale, ...rest].join("/")
```
Then next-intl will handle this as no region whatsoever and it won't need to have it as a `[country]` in your structure. 

Though, you'll have to work around `Link` component, we ended up writing our own. But frankly, it was the least problem. 

Sorry I can't give a more complete solution :c

**Reply by amannn** [reply]
Seems like a doable solution, yes.

Just for the record, [custom prefixes](https://next-intl-docs.vercel.app/docs/routing#locale-prefix-custom) won't work for your use case? E.g. `/es/es-mx` → `/es-MX`? Your redirects that handle the case for `/[country]` → `/[country]/[locale]` can still be used on top.

As a side note, the Next.js team is working on [`rootParams`](https://github.com/vercel/next.js/pull/72837) which I think will allow for a cleaner solution to this situation once it's available :).

**Reply by HappyEmu** [reply]
@amannn 
Yeah that looks promising, hope they get that into stable soon 🙏 . Seems like this is necessary to properly handle arbitrary params in front of a [locale] segment.
And yes, if I understood correctly, in my case I cannot use custom prefixes. I will need access to [country] and my country-locale combinations are arbitrary (not adhering to Intl standards, e.g. CH-en), so I couldn't infer them (like in the example here https://next-intl-docs.vercel.app/docs/routing#locale-prefix-custom-read-prefix)

**Reply by amannn** [reply]
> I will need access to [country] and my country-locale combinations are arbitrary (not adhering to Intl standards, e.g. CH-en), so I couldn't infer them

Just to make sure we're on the same page: Custom prefixes only require that _internal_ pathnames use a valid locale, the _external_ ones can be completely arbitrary.

On top of that, you can use private extensions on a locale to encode arbitrary information:

> If you’d like to encode custom information in the locale, you can use arbitrary [private extensions](https://tc39.es/proposal-intl-locale/#sec-insert-unicode-extension-and-canonicalize), denoted by the -x- prefix (e.g. en-US-x-usd).

My current understanding is that the only use case that's currently not supported with custom prefixes is when you have an unknown, infinite number of arbitrary prefixes. For other cases I think you should have quite some flexibility.

Is that the case in your project? Or am I missing something?

**Reply by HappyEmu** [reply]
No, you are right, I don't have an infinite number of arbitrary prefixes. Regrettably, I stopped reading before the private extensions paragraph, which would presumably solve my problem. I just need to be able to extract the country from the `locale` that is exposed by `useLocale()` and I should be able to do that using the private extensions 👍 

Thank you so much for your help 🙏 

Just from a conceptual point of view, it would still be much nicer to have the routes actually stay [country]/[locale]. If that NextJS PR could enable that, I'm all for it 😄

**Reply by amannn** [reply]
That sounds great! 🙌

> Just from a conceptual point of view, it would still be much nicer to have the routes actually stay [country]/[locale]. If that NextJS PR could enable that, I'm all for it 😄

Yep, I agree!

I also mentioned this discussion in https://github.com/amannn/next-intl/issues/1107, these seem closely related.

**Reply by Maclay74** [reply]
> Just for the record, [custom prefixes](https://next-intl-docs.vercel.app/docs/routing#locale-prefix-custom) won't work for your use case? E.g. `/es/es-mx` → `/es-MX`? Your redirects that handle the case for `/[country]` → `/[country]/[locale]` can still be used on top.

Yea, we tried that. We looped over regions and languages generating those prefixes. It worked, but poorly. Our webserver couldn't process so many alternative links in header, so we had to disable them and other stuff. But the main issue with this approach is that there is no way to control what locale is passed to `Intl`. And if it's `es-mx-MX` or something like that, than it throws an error. 

> As a side note, the Next.js team is working on https://github.com/vercel/next.js/pull/72837 which I think will allow for a cleaner solution to this situation once it's available :).

Oh, haven't seen this one! Cool!
